### PR TITLE
better implementation for store subscription without prototypal inheritance

### PR DESCRIPTION
**Pull request by Ma27**
#### Changes

This is a new approach for the `subscribeStores` implementation using "real" higher order components similar to the ones implemented in other flux-like libraries like `react-redux` or `yahoo/fluxible`.
We generate a new component surrounding the one to decorate to fulfill with "composition over inheritance" as inheritance as done in 1.0.0 is quite dangerous and the new approach is able to deal with stateless components as everything is passed to the component via `props`.
##### Testing

The downside of the new approach is that `shallow` rendering is harder because it renders the first level of the JSX tree starting from a certain component only. If we render the exported component, the result of the generated component could be tested, but not the component itself.

The current solution is a function called `pure` which retrieves the component from this automatically generated tree.

Testing could look like this:

``` javascript
import { pure } from 'sententiaregum-flux-container';
import { shallow } from 'enzyme';
import example from '../../components/example';
import { expect } from 'chai';

describe('example', () => {
  it('renders a headling', () => {
    // Tree of the actual `example` component:
    // render() {
    //   return (
    //     <h1>{this.props.headline}</h1>
    //   );
    // }
    const tree = shallow(pure(example, { headline: 'Foo' }));

     expect(tree.text(), 'Foo');
  });
});
```
##### Validate whether to re-render

Some libraries check whether something has changed before running a `setState` which causes a re-creation of the tree and likely a re-render.
Sometimes this can be harmful even if no "real" changes were done in the store. In our case this is not definetly necessary as a store just flushes a change event to the view if something has "really" changed which will be validated using a `deep-equal` check (https://github.com/Sententiaregum/flux-container/blob/master/src/store.js#L53)
##### support for v1

As this will land in 2.0, I've created a simple support plan for 1.0:

Even though it's not sure whether anyone uses that right now, I'd support 1.0 as it's not a hard impact and I always checkout branches for a minor release to simplify maintenance.
### Breaking changes

This should land in `2.0.0` because of the following reasons:
- API changes in the `subscribeStores`
- the result is different (impact when doing `shallow` testing with `enzyme`)

### CODE REVIEW

File: src/util/pure.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * Components itself might be pure, but are modified by using the `subscribeStores` function. However, this makes
+ * testing, especially `shallow rendering` more difficult, so a helper API
+ * which provides the required access is necessary.
+ *
+ * @param {Object} connectedParent The connected parent which will be generated by the `subscribeStores` function.
+ * @param {Object} props           The props for the function to test.
+ *
+ * @returns {React.Element} A newly created react.js element.
+ */
+export default (connectedParent, props) => React.createElement(connectedParent.wrapped(), props);
```
Comment (Ma27) [inline src/util/pure.js]: we should check whether the `wrapped()` method exists and throw a better message after that. This isn't always the case, if a component is not decorated.

File: test/indexSpec.js [hunk 2]
```diff
@@ -10,109 +10,39 @@
...
-
-        // log execution times
-        testStub();
-      }
-      modify(newText) {
-        this.setState({ text: newText });
-      }
-    }, {
-      'refresh': testStore
+    const component = props => <h1>{props.header}</h1>;
```
Comment (Ma27) [inline test/indexSpec.js:43]: might be useful to add another test with a stateful component

**Review by Ma27 — COMMENTED** [review]
minor things to fix to avoid odd regressions, but no harmful things.

**Review by Ma27 — COMMENTED** [review]
looks good for now. @benbieler would you approve?

**Review by benbieler — APPROVED** [review]
Fine :+1:

**Pull request merged**